src/ucp/tag/eager_rcv.c

```diff
@@ -15,65 +15,61 @@
+ucp_eager_tagged_handler(void *arg, void *data, size_t length, unsigned am_flags,
+                         uint16_t flags, uint16_t hdr_len)
 {
-    ucp_worker_h worker                      = arg;
-    ucp_eager_hdr_t *eager_hdr               = data;
-    ucp_eager_first_hdr_t *eager_first_hdr   = data;
-    ucp_eager_middle_hdr_t *eager_middle_hdr = data;
+    ucp_worker_h worker        = arg;
+    ucp_eager_hdr_t *eager_hdr = data;
+    ucp_eager_first_hdr_t *eagerf_hdr;
+    ucp_recv_desc_t *rdesc;
     ucp_request_t *req;
     ucs_status_t status;
-    size_t recv_len;
     ucp_tag_t recv_tag;
-    ucp_recv_desc_t *rdesc;
+    size_t recv_len;
 
-    UCS_PROFILE_SCOPE_BEGIN();
     ucs_assert(length >= hdr_len);
-    /* TODO: split packets: shift middle packets to separate queue */
-    recv_tag = (flags & UCP_RECV_DESC_FLAG_MIDDLE) ?
-               eager_middle_hdr->super.tag : eager_hdr->super.tag;
+    ucs_assert(flags & UCP_RECV_DESC_FLAG_EAGER);
+
+    recv_tag = eager_hdr->super.tag;
     recv_len = length - hdr_len;
 
-    req = ucp_tag_exp_search(&worker->tm, recv_tag, recv_len, flags);
+    req = ucp_tag_exp_search(&worker->tm, recv_tag);
     if (req != NULL) {
+        ucs_trace_req("found req %p", req);
         UCS_PROFILE_REQUEST_EVENT(req, "eager_recv", recv_len);
 
-        status = ucp_request_recv_data_unpack(req, data + hdr_len, recv_len,
-                                              req->recv.state.offset,
-                                              flags & UCP_RECV_DESC_FLAG_LAST);
-
         /* First fragment fills the receive information */
-        if (flags & UCP_RECV_DESC_FLAG_FIRST) {
-            UCP_WORKER_STAT_EAGER_MSG(worker, flags);
-            req->recv.tag.info.sender_tag = recv_tag;
-
-            /* Cancel req in transport if it was offloaded,
-             * because it arrived either:
-             * 1) via SW TM (e. g. peer doesn't support offload)
-             * 2) as unexpected via HW TM */
-            ucp_tag_offload_try_cancel(worker, req, 1);
-
-            if (flags & UCP_RECV_DESC_FLAG_LAST) {
-                req->recv.tag.info.length = recv_len;
-            } else {
-                req->recv.tag.info.length = eager_first_hdr->total_len;
-            }
+        UCP_WORKER_STAT_EAGER_MSG(worker, flags);
+        UCP_WORKER_STAT_EAGER_CHUNK(worker, EXP);
+
+        req->recv.tag.info.sender_tag = recv_tag;
+
+        /* Cancel req in transport if it was offloaded,
+         * because it arrived either:
+         * 1) via SW TM (e. g. peer doesn't support offload)
+         * 2) as unexpected via HW TM */
+        ucp_tag_offload_try_cancel(worker, req, 1);
+
+        if (flags & UCP_RECV_DESC_FLAG_EAGER_SYNC) {
+            ucp_tag_eager_sync_send_ack(worker, data, flags);
         }
 
-        /* Last fragment completes the request */
-        if (flags & UCP_RECV_DESC_FLAG_LAST) {
+        if (flags & UCP_RECV_DESC_FLAG_EAGER_ONLY) {
```

Comment:
what about make decision about complete message (EAGER_ONLY here) based on message == received size?

Reply:
not sure what you mean here.. but EAGER_ONLY header does not contain a total message size field

Reply:
right, I mean add length to EAGER_ONLY packet & unify all messages to FIRST/MIDDLE channels, but it is topic for discussions

Reply:
i'd rather not add another field to eager_only, since it's used for very small messages as well, and it will add extra overhead